### ISSUE TITLE
how to set maxBackoff value 

### ISSUE THREAD

**Issue by louie-wh** [grpc/grpc-java#10932]
I want to set a small value maxBackoff,  because I want quickly to reconnect to service. 
But default reconnect policy is maxBackoff will increased

so I want to use defaultServiceConfig to control maxBackoff of channel.

```kotlin
    private fun getHmiServiceChannel():ManagedChannel{
        val ip = RepositoryDataStore.getValueSync(KEY_DATA_STORE_GRPC_IP, HOST_DV_SERVICE)
        initGprcDebugLog()

        // keepAliveTime: 表示当grpc连接没有数据传递时，多久之后开始向server发送ping packet
        // keepAliveTimeout: 表示当发送完ping packet后多久没收到server回应算超时
        // keepAliveTimeoutCalls: 表示如果grpc连接没有数据传递时，是否keepAlive,默认为false
        return ManagedChannelBuilder
            .forAddress(ip, PORT_DV_SERVICE)
            .usePlaintext()
            .keepAliveTime(5, TimeUnit.SECONDS)
            .keepAliveTimeout(1,TimeUnit.SECONDS)
            .keepAliveWithoutCalls(true)
            .enableRetry()
            .maxRetryAttempts(10000)
            .defaultServiceConfig(getRetryingServiceConfig())
            .intercept(ClientLogInterceptor())
            .build()
    }


    private fun getRetryingServiceConfig(): MutableMap<String, *> {
        val config = """
            {
                "methodConfig":[
                    {
                        "name":[
                            {
                                "service":"interface.proto.dreamview.HMIService",
                                "method":"WatchVehicleStatus"
                            }
                        ],
                        "retryPolicy":{
                            "maxAttempts":5,
                            "initialBackoff":"0.5s",
                            "maxBackoff":"5s",
                            "backoffMultiplier":1,
                            "retryableStatusCodes":[
                                "UNAVAILABLE"
                            ]
                        }
                    }
                ]
            }
        """.trimIndent()
        return Gson().fromJson(config, MutableMap::class.java) as MutableMap<String, *>
    }
```

but no effect

**Comment by louie-wh** [comment]
this is a retry log 

2024-02-20 17:03:37.073 18758-18935 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Subchannel<3>: (10.2.5.182:8421)] CONNECTING after backoff
2024-02-20 17:03:37.081 18758-18935 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Subchannel<3>: (10.2.5.182:8421)] Started transport OkHttpClientTransport<21>: (/10.2.5.182:8421)
2024-02-20 17:03:37.093 18758-20332 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Subchannel<3>: (10.2.5.182:8421)] OkHttpClientTransport<21>: (/10.2.5.182:8421) SHUTDOWN with UNAVAILABLE[java.net.ConnectException: failed to connect to /10.2.5.182 (port 8421) from /:: (port 50928): connect failed: ECONNREFUSED (Connection refused)]
2024-02-20 17:03:37.097 18758-20332 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Subchannel<3>: (10.2.5.182:8421)] TRANSIENT_FAILURE (UNAVAILABLE[java.net.ConnectException: failed to connect to /10.2.5.182 (port 8421) from /:: (port 50928): connect failed: ECONNREFUSED (Connection refused)]). Will reconnect after 107,567,755,237 ns
2024-02-20 17:03:37.099 18758-20332 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Channel<1>: (10.2.5.182:8421)] Entering TRANSIENT_FAILURE state with picker: Picker{result=PickResult{subchannel=null, streamTracerFactory=null, status=Status{code=UNAVAILABLE, description=null, cause=java.net.ConnectException: failed to connect to /10.2.5.182 (port 8421) from /:: (port 50928): connect failed: ECONNREFUSED (Connection refused)
                                                                                                    	at libcore.io.IoBridge.connect(IoBridge.java:188)
                                                                                                    	at java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:142)
                                                                                                    	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:390)
                                                                                                    	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:230)
                                                                                                    	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:212)
                                                                                                    	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:436)
                                                                                                    	at java.net.Socket.connect(Socket.java:621)
                                                                                                    	at java.net.Socket.connect(Socket.java:570)
                                                                                                    	at java.net.Socket.<init>(Socket.java:450)
                                                                                                    	at java.net.Socket.<init>(Socket.java:250)
                                                                                                    	at javax.net.DefaultSocketFactory.createSocket(SocketFactory.java:285)
                                                                                                    	at io.grpc.okhttp.OkHttpClientTransport$3.run(OkHttpClientTransport.java:535)
                                                                                                    	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
                                                                                                    	at java.lang.Thread.run(Thread.java:933)
                                                                                                    Caused by: android.system.ErrnoException: connect failed: ECONNREFUSED (Connection refused)
                                                                                                    	at libcore.io.Linux.connect(Native Method)
                                                                                                    	at libcore.io.ForwardingOs.connect(ForwardingOs.java:204)
                                                                                                    	at libcore.io.BlockGuardOs.connect(BlockGuardOs.java:157)
                                                                                                    	at libcore.io.ForwardingOs.connect(ForwardingOs.java:204)
                                                                                                    	at libcore.io.IoBridge.connectErrno(IoBridge.java:202)
                                                                                                    	at libcore.io.IoBridge.connect(IoBridge.java:180)
                                                                                                    	at java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:142) 
                                                                                                    	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:390) 
                                                                                                    	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:230) 
                                                                                                    	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:212) 
                                                                                                    	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:436) 
                                                                                                    	at java.net.Socket.connect(Socket.java:621) 
                                                                                                    	at java.net.Socket.connect(Socket.java:570) 
                                                                                                    	at java.net.Socket.<init>(Socket.java:450) 
                                                                                                    	at java.net.Socket.<init>(Socket.java:250) 
                                                                                                    	at javax.net.DefaultSocketFactory.createSocket(SocketFactory.java:285) 
                                                                                                    	at io.grpc.okhttp.OkHttpClientTransport$3.run(OkHttpClientTransport.java:535) 
                                                                                                    	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133) 
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
                                                                                                    	at java.lang.Thread.run(Thread.java:933) 
                                                                                                    }, drop=false}}
2024-02-20 17:03:37.100 18758-20332 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Channel<1>: (10.2.5.182:8421)] Resolved address: [[[/10.2.5.182:8421]/{}]], config={io.grpc.internal.RetryingNameResolver.RESOLUTION_RESULT_LISTENER_KEY=io.grpc.internal.RetryingNameResolver$ResolutionResultListener@4d1f815}
2024-02-20 17:03:37.100 18758-20332 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Channel<1>: (10.2.5.182:8421)] Received no service config, using default service config
2024-02-20 17:03:37.103 18758-20334 GrpcService             tech.autra.auto                      D  io.grpc.ChannelLogger   log   [Subchannel<3>: (10.2.5.182:8421)] OkHttpClientTransport<21>: (/10.2.5.182:8421) Terminated
2024-02-20 17:03:37.431 18758-18870 GrpcService             tech.autra.auto                      I  handleMessage -> { when=-6ms realwhen=3548770 isVsync=false isAsynchronous=false what=2 target=tech.autra.auto.repository.grpc.GrpcService$keepAliveHandler$1 }
2024-02-20 17:03:37.431 18758-18870 GrpcService             tech.autra.auto                      I  gr

**Comment by louie-wh** [comment]
Thanks
[resetConnectBackoff](https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannel.html#resetConnectBackoff())  work good.

I also try notifyWhenStateChanged to listener the state of channel .
but  notifyWhenStateChanged is once call, and the parameter of ConnectivityState is complex.  so I using getState  at a new thread loop to get the state of channel. 

and this issue   [](https://github.com/grpc/grpc-java/issues/4359) show why  notifyWhenStateChanged designed .

I still think channel should have a state observer API. and Application deal with the rapidly state change.
